### ISSUE TITLE
Promise resolution and reference

### ISSUE THREAD

**Issue by bpolaszek** [guzzle/promises#57]
Hi everyone,

## TL; DR
Make promise resolution simpler:
```php
$promise = new \GuzzleHttp\Promise\Promise(function () use (&$promise) {
    $promise->resolve('The referenced promise has been resolved');
});
```

to:
```php
$promise = new \GuzzleHttp\Promise\Promise(function (PromiseInterface $promise) {
    $promise->resolve('The promise has been resolved');
});
```

## Example

Actually, when you create a promise from scratch you **must** use a _reference to that promise_ within the wait function, i.e.:

```php
$promise = new \GuzzleHttp\Promise\Promise(function () use (&$promise) {
    $promise->resolve('The referenced promise has been resolved');
});
```

From a DX point of view, this can lead to headaches when it comes to generate promises within a loop. Let's take this as an example:

```php
use GuzzleHttp\Promise\Promise;
use function GuzzleHttp\Promise\all;

require_once __DIR__ . '/vendor/autoload.php';

$translationMap = [
    'banana' => 'banane',
    'apple'  => 'pomme',
];

$fruits = [
    'banana',
];

$promises = [];

foreach ($fruits AS $fruit) {
    $promises[$fruit] = $promise = new Promise(function () use (&$promise, $translationMap, $fruit) {
        $promise->resolve($translationMap[$fruit]);
    });
}

var_dump(all($promises)->wait()); // Expected ['banana' => 'banane'] and indeed got ['banana' => 'banane']
```

In this example, there's only 1 instance of `$promise`, and everything's fine. Now add an `apple` into the `$fruits` array:

```php
use GuzzleHttp\Promise\Promise;
use function GuzzleHttp\Promise\all;

require_once __DIR__ . '/vendor/autoload.php';

$translationMap = [
    'banana' => 'banane',
    'apple'  => 'pomme',
];

$fruits = [
    'banana',
    'apple',
];

$promises = [];

foreach ($fruits AS $fruit) {
    $promises[$fruit] = $promise = new Promise(function () use (&$promise, $translationMap, $fruit) {
        $promise->resolve($translationMap[$fruit]);
    });
}

var_dump(all($promises)->wait()); // Expected ['banana' => 'banane', 'apple' => 'pomme'] 
// But got GuzzleHttp\Promise\RejectionException: Invoking the wait callback did not resolve the promise...
```

Of course, this `RejectionException` would not happen if I chose a different variable name for each `$promise`. But honestly this is ugly.

A simpler solution would be to pass the promise itself directly as an argument into the wait function, instead of relying on a reference, i.e:
```php
$promise = new \GuzzleHttp\Promise\Promise(function (PromiseInterface $promise) {
    $promise->resolve('The promise has been resolved');
});
```

This would require a change into the `Promise::invokeWaitFn()` method:
```php
$wfn($this);
// instead of
$wfn(true);
```

This would simplify a lot the promise resolution.
What's your opinion ?

**Comment by bpolaszek** [comment]
Looks like my post has reached a black hole.
![Forever Alone](https://images.encyclopediadramatica.se/0/0e/Forever_Alone.png)

**Comment by mtdowling** [comment]
> Actually, when you create a promise from scratch you must use a reference to that promise within the wait function

This isn't true if the method in which you block (i.e., wait) is to block on an event loop or something like it. For example, in Guzzle's curl handlers, waiting on a promise just ticks the curl event loop. Doing so should cause the outstanding promise to complete. From what I've seen, most cases of needing to pass a reference to a promise wait function is for examples or tests.

That said, I see that it could be useful for some implementations that want to share a wait function between promises and have the specific promise instance passed into the function (e.g., this could be used to more easily reason about how or what to block on).

So, sure, I think this sounds fine to me. I don't see how it could break any code in the wild.

@sagikazarmark do you have any thoughts on this?

**Comment by bpolaszek** [comment]
Hi @mtdowling, thank you for replying.
I actually have a full app relying on Promises everywhere, so sometimes my promises are just returned by Guzzle HTTP Client, sometimes they're constructed by myself.
For instance, I can code a `wait()` function that will wait for a separate PHP process to complete, or a wrapper for guzzle promises to avoir querying a remote and heavy API if a result is already stored in Redis.

Here is a sample code that makes use of a Promise constructor in a loop:

```php
use GuzzleHttp\Promise\PromiseInterface;
interface ApiInterface {

    /**
     * Fetch heavy data on a very slow remote API
     * @return PromiseInterface
     */
    public function fetchDataAsync(): PromiseInterface;
}
```

This example implementation will call 2 APIs to `fetchDataAsync` - This will take ~ 8 seconds.
```php
use GuzzleHttp\Client;
use GuzzleHttp\Promise\Promise;
use GuzzleHttp\Promise\PromiseInterface;
use GuzzleHttp\Psr7\Request;
use Psr\Cache\CacheItemInterface;
use Psr\Cache\CacheItemPoolInterface;
use Psr\Http\Message\ResponseInterface;
use Psr\Log\LoggerInterface;
use function GuzzleHttp\Promise\all;

class MyApiSample implements ApiInterface {

    const API_URLS = [
        'bar' => 'http://httpbin.org/delay/3?nb_clicks=100',
        'baz' => 'http://httpbin.org/delay/5?nb_clicks=150',
    ];

    /**
     * @var ClientInterface
     */
    private $client;

    /**
     * @var CacheItemPoolInterface
     */
    private $cachePool;

    /**
     * @var LoggerInterface
     */
    private $logger;

    /**
     * MyApiSample constructor.
     * @param ClientInterface       $client
     * @param CacheItemPoolInterface $cachePool
     * @param LoggerInterface          $logger
     */
    public function __construct(ClientInterface $client,
                                CacheItemPoolInterface $cachePool,
                                LoggerInterface $logger) {
        $this->client    = $client;
        $this->cachePool = $cachePool;
        $this->logger    = $logger;
    }

    /**
     * @inheritDoc
     */
    public function fetchDataAsync(): PromiseInterface {

        $promises = new ArrayIterator();

        foreach (static::API_URLS AS $key => $url) {

            // Note that the promise is injected into the waitFn constructor.
            $promises[$key] = new Promise(function (PromiseInterface $promise) use ($key, $url) {

                $cacheItem = $this->cachePool->getItem($key);

                if (!$cacheItem->isHit()) {
                    $this->logger->info(sprintf('%s - MISS', $url));
                    $promise->resolve($this->fetchRemoteDataAsync($url)->then(function (array $json) use ($cacheItem) {
                        $this->store($cacheItem, $json);
                        return $json;
                    }));
                }
                else {
                    $this->logger->info(sprintf('%s - HIT', $url));
                    $promise->resolve($cacheItem->get());
                }

            });

        }

        return all($promises)->then(function (array $results) {
            return array_sum(array_map(function (array $result) {
                return $result['args']['nb_clicks'];
            }, $results));

        });
    }

    /**
     * Really calls the API.
     * @param $url
     * @return PromiseInterface
     */
    private function fetchRemoteDataAsync($url): PromiseInterface {
        return $this->client->sendAsync(new Request('GET', $url))->then(function (ResponseInterface $response) {
            return $this->read($response);
        });
    }

    /**
     * Transforms the HTTP response into a data array.
     * @param ResponseInterface $response
     * @return array json decoded data
     */
    private function read(ResponseInterface $response): array {
        return json_decode((string) $response->getBody(), true);
    }

    /**
     * Stores data into the cache pool.
     * @param CacheItemInterface $cacheItem
     * @param array                         $json
     * @param int                           $ttl
     */
    private function store(CacheItemInterface $cacheItem, array $json, $ttl = 300) {
        $cacheItem->set($json);
        $cacheItem->expiresAfter($ttl);
        $this->cachePool->save($cacheItem);
    }
}

```

Now we'll call `fetchDataAsync` twice. Instead of 16 seconds it will only take 8.
```php
use GuzzleHttp\Client;
use Monolog\Handler\TestHandler;
use Monolog\Logger;
use Symfony\Component\Cache\Adapter\ArrayAdapter;
$start      = time();
$guzzle     = new Client();
$cachePool  = new ArrayAdapter();
$logger     = new Logger('api');
$logHandler = new TestHandler();
$logger->pushHandler($logHandler);
$myApiSample = new MyApiSample($guzzle, $cachePool, $logger);

printf('Number of clicks is %d' . PHP_EOL, $myApiSample->fetchDataAsync()->wait());
foreach ($logHandler->getRecords() AS $record) {
    echo $record['message'] . PHP_EOL;
}

# For some reason I have to `fetchDataAsync()` a few moments later - then I'll hit the cache
$logHandler->clear();
echo PHP_EOL;
printf('Number of clicks is %d' . PHP_EOL, $myApiSample->fetchDataAsync()->wait());
foreach ($logHandler->getRecords() AS $record) {
    echo $record['message'] . PHP_EOL;
}

$end = time();
echo PHP_EOL;
printf('It took %d seconds to process.', round($end - $start));
```

Here's the output:
```

Number of clicks is 250
http://httpbin.org/delay/3?nb_clicks=100 - MISS
http://httpbin.org/delay/5?nb_clicks=150 - MISS

Number of clicks is 250
http://httpbin.org/delay/3?nb_clicks=100 - HIT
http://httpbin.org/delay/5?nb_clicks=150 - HIT

It took 8 seconds to process.
```

In my real use case, I don't rely on an `ArrayAdapter` but on a `RedisAdapter`. This means multiple PHP processes can call `fetchDataAsync` and may hit the cache at the first time.

**Comment by venkatsamuthiram** [comment]
guzzlehttp/guzzle

$promise  = $client->sendAsync($request);
            //You can force an asynchronous response to complete using the wait() method of the returned promise
            $response = $promise->wait();
            $promiseResult     = $promise->then(function ($response) {
                $responseArray = [];
                $responseArray += ['statusCode' => $response->getStatusCode()];
                $responseArray += ['reasonPhrase' => $response->getReasonPhrase()];
                $responseArray += ['body' => json_decode($response->getBody())];
                $result        =  json_encode($responseArray);
                var_dump($result);
                return $result;
            });
This is my Code, How? I return my Response data?

**Comment by Tobion** [comment]
The problem I see with the proposal 

```
$promise = new \GuzzleHttp\Promise\Promise(function (PromiseInterface $promise) {
    $promise->resolve('The promise has been resolved');
});
```

is that it not clear anymore that the passed promise argument is the same promise as the outer one. So in theory you don't know which promise you are resolving. That does not seem right to me.

**Comment by cocochepeau** [comment]
Why not doing like in JS?

```php
$promise = new \GuzzleHttp\Promise\Promise(function ($resolve, $reject) {
    $resolve('The promise has been resolved');
});
```
